controllers/remote/cluster_cache.go

```diff
@@ -0,0 +1,401 @@
+	m.watchesLock.RUnlock()
+
+	if exists {
+		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
+		return nil
+	}
+
+	// Doesn't exist - grab the write lock
+	m.watchesLock.Lock()
+	defer m.watchesLock.Unlock()
+
+	// Need to check if another goroutine created the watch while this one was waiting for the lock
+	if m.watchExists(input.Cluster, wi) {
+		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
+		return nil
+	}
+
+	// Need to create the watch
+	watchesForCluster, found := m.watches[input.Cluster]
+	if !found {
+		watchesForCluster = make(map[watchInfo]struct{})
+		m.watches[input.Cluster] = watchesForCluster
+	}
+
+	cache, err := m.getOrCreateClusterCache(ctx, input.Cluster)
+	if err != nil {
+		return err
+	}
+
+	if err := input.Watcher.Watch(source.NewKindWithCache(input.Kind, cache), input.EventHandler, input.Predicates...); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("error creating watch"))
+	}
+
+	watchesForCluster[wi] = struct{}{}
+
+	return nil
+}
+
+// getOrCreateClusterCache returns the clusterCache for cluster, creating a new ClusterCache if needed.
+func (m *ClusterCacheTracker) getOrCreateClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
+	cache := m.getClusterCache(cluster)
+	if cache != nil {
+		return cache, nil
+	}
+
+	return m.newClusterCache(ctx, cluster)
+}
+
+// getClusterCache returns the clusterCache for cluster, or nil if it does not exist.
+func (m *ClusterCacheTracker) getClusterCache(cluster client.ObjectKey) *clusterCache {
+	m.clusterCachesLock.RLock()
+	defer m.clusterCachesLock.RUnlock()
+
+	return m.clusterCaches[cluster]
+}
+
+// newClusterCache creates and starts a new clusterCache for cluster.
+func (m *ClusterCacheTracker) newClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
+	m.clusterCachesLock.Lock()
+	defer m.clusterCachesLock.Unlock()
+
+	// If another goroutine created the cache while this one was waiting to acquire the write lock, return that
+	// instead of overwriting it.
+	if c, exists := m.clusterCaches[cluster]; exists {
+		return c, nil
+	}
+
+	config, err := RESTConfig(ctx, m.client, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
+	}
+
+	remoteCache, err := cache.New(config, cache.Options{Scheme: m.scheme})
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating cache for remote cluster")
+	}
+	stop := make(chan struct{})
+
+	cc := &clusterCache{
+		Cache: remoteCache,
+		stop:  stop,
+	}
+	m.clusterCaches[cluster] = cc
+
+	// Start the cache!!!
+	go remoteCache.Start(cc.stop)
+	// Start cluster healthcheck!!!
+	go m.healthCheckCluster(ctx, cluster, config, healthCheckPollInterval, healthCheckRequestTimeout, healthCheckUnhealthyThreshold, "/")
```

Comment:
I don't think you can use `ctx` here as a means to stop this goroutine. `ctx` is passed in by a controller that calls `ClusterCacheTracker.Watch()`, which presumably is a pre-reconcile context.

Could we tie its lifecycle to the remote cache's `cc.stop`?